### ISSUE TITLE
[distributed_c10d] Provide option in barrier function to pass rank to GPU mapping

### ISSUE THREAD

**Issue by jaglinux** [pytorch/pytorch#48110]
## 🐛 Bug
As part of #47629, barrier() in distributed_c10d.py should provide option to pass the tensor or rank to GPU mapping.
This will enable the test cases to use the same GPU ranking while calling default pg tests / global tests.
This bug is specific to NCCL case.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd

**Comment by jaglinux** [comment]
cc @jeffdaily

**Comment by mrshenli** [comment]
Hey @jaglinux, could you please provide some more details? E.g., if we add this argument to `barrier()`, what the format should be, and how we should explain that to users?

**Comment by jaglinux** [comment]
The format can be  barrier(group=group.WORLD,async_op=False, tensor=None) similar to broadcast() which provides argument to pass in tensor. 
https://github.com/pytorch/pytorch/blob/master/torch/distributed/distributed_c10d.py#L2176

User can pass the tensor argument (if required) with user defined rank to GPU mapping. The tensors will be used in the allreduce operation inside barrier definition for NCCL backend.
In that case; init_process_group, new_group(), _new_process_group_helper() should be modified to allow the tensor argument in addition to ProcessGroupNCCL.cpp and other cpp file changes.

**Comment by jaglinux** [comment]
I think adding tensor parameter to barrier is taxing for user since its not user intention to call all_reduce inside barrier.
Instead of that, adding new parameter "device" is meaningful where device is the preferred GPU to be used for a particular rank.
def barrier(group=group.WORLD,async_op=False, device=None)
The below functions also needs to be changed to accommodate GPU preference.
def init_process_group(backend, ... ,device=None): 
def new_group(ranks=None, timeout=default_pg_timeout, backend=None, device=None)

This is inline with the example at https://pytorch.org/tutorials/intermediate/ddp_tutorial.html
For demo_basic case, if the user wants to assign different device to the rank
        # initialize the process group
        dist.init_process_group("gloo", rank=rank, world_size=world_size, **device=rank*2**)
And then change the device preferences in demo_basic() function accordingly.

Please note that in the current form, the demo_basic fails if different rank to GPU mapping is used,
To reproduce the issue,
@@ -49,15 +49,16 @@ def demo_basic(rank, world_size):
     setup(rank, world_size)

     # create model and move it to GPU with id rank
-    model = ToyModel().to(rank)
-    ddp_model = DDP(model, device_ids=[rank])
+    # use a different rank to GPU mapping
+    model = ToyModel().to(rank*2)
+    ddp_model = DDP(model, device_ids=[rank*2])

     loss_fn = nn.MSELoss()
     optimizer = optim.SGD(ddp_model.parameters(), lr=0.001)

     optimizer.zero_grad()
     outputs = ddp_model(torch.randn(20, 10))
-    labels = torch.randn(20, 5).to(rank)
+    labels = torch.randn(20, 5).to(rank*2)
     loss_fn(outputs, labels).backward()
     optimizer.step()

@@ -75,6 +76,6 @@ if __name__ == "__main__":
     if n_gpus < 8:
       print(f"Requires at least 8 GPUs to run, but got {n_gpus}.")
     else:
-      run_demo(demo_basic, 8)
+      run_demo(demo_basic, 2)
       run_demo(demo_checkpoint, 8)
       run_demo(demo_model_parallel, 4)

**Comment by jaglinux** [comment]
@mrshenli any review comments for the above PR ?

**Comment by mrshenli** [comment]
Hey @jaglinux, thanks for sharing more details and sorry about the delay. Let's discuss on the PR.
